test/extensions/filters/listener/common/fuzz/listener_filter_fuzzer.cc

```diff
@@ -10,19 +8,85 @@ void ListenerFilterFuzzer::fuzz(
     const test::extensions::filters::listener::FilterFuzzTestCase& input) {
   try {
-    fuzzerSetup(input);
+    socket_.setLocalAddress(Network::Utility::resolveUrl(input.sock().local_address()));
+  } catch (const EnvoyException& e) {
+    // If fuzzed local address is malformed or missing, socket's local address will be nullptr
+  }
+  try {
+    socket_.setRemoteAddress(Network::Utility::resolveUrl(input.sock().remote_address()));
   } catch (const EnvoyException& e) {
-    ENVOY_LOG_MISC(debug, "EnvoyException: {}", e.what());
-    return;
+    // If fuzzed remote address is malformed or missing, socket's remote address will be nullptr
+  }
+
+  const int nreads = input.data_size(); // Number of reads
+
+  if (nreads > 0) {
+    EXPECT_CALL(socket_, detectedTransportProtocol()).WillRepeatedly(testing::Return("raw_buffer"));
+
+    EXPECT_CALL(os_sys_calls_, recv(42, _, _, MSG_PEEK))
+        .WillOnce(testing::Return(Api::SysCallSizeResult{static_cast<ssize_t>(0), 0}));
+
+    EXPECT_CALL(dispatcher_,
+                createFileEvent_(_, _, Event::FileTriggerType::Edge,
+                                 Event::FileReadyType::Read | Event::FileReadyType::Closed))
+        .WillOnce(testing::DoAll(testing::SaveArg<1>(&file_event_callback_),
+                                 testing::ReturnNew<NiceMock<Event::MockFileEvent>>()));
   }
 
   filter.onAccept(cb_);
-}
 
-void ListenerFilterFuzzer::socketSetup(
-    const test::extensions::filters::listener::FilterFuzzTestCase& input) {
-  socket_.setLocalAddress(Network::Utility::resolveUrl(input.sock().local_address()));
-  socket_.setRemoteAddress(Network::Utility::resolveUrl(input.sock().remote_address()));
+  if (nreads > 0) {
+    // Construct header from single or multiple reads
+    std::string data = "";
+    std::vector<size_t> indices; // Ending indices for each read
+
+    size_t curr = 0;
+
+    for (int i = 0; i < nreads; i++) {
+      data += input.data(i);
+      curr += input.data(i).size();
+      indices.push_back(curr);
+    }
+
+    absl::string_view header(data);
+
+    int nread = 0; // Counter of current read
+
+    {
+      testing::InSequence s;
+
+      if (nreads > 1) {
+        EXPECT_CALL(os_sys_calls_, recv(42, _, _, MSG_PEEK))
+            .WillOnce(testing::InvokeWithoutArgs([]() {
+              return Api::SysCallSizeResult{ssize_t(-1), SOCKET_ERROR_AGAIN};
+            }));
+      }
+
+      EXPECT_CALL(os_sys_calls_, recv(42, _, _, MSG_PEEK))
+          .Times(testing::AnyNumber())
+          .WillRepeatedly(Invoke([&header, &indices, &nread](os_fd_t, void* buffer, size_t length,
+                                                             int) -> Api::SysCallSizeResult {
+            ASSERT(length >= indices[nread]);
+            memcpy(buffer, header.data(), indices[nread]);
+            return Api::SysCallSizeResult{ssize_t(indices[nread++]), 0};
+          }));
+    }
+
+    bool got_continue = false;
+
+    ON_CALL(cb_, continueFilterChain(true))
```

Comment:
Is this ON_CALL necessary to do in every execution? Is it possible to make this `ListenerFilterFuzzer` object `static` in each fuzz target(such as in `http_inspector_fuzz_test.cc`) and  setup some `ON_CALL.ByDefault()` only for one time in `ListenerFilterFuzzer`'s constructor?

Reply:
I think so, since the lambda sets got_continue, which needs to be reset every time fuzz() is called. There might be a way to implement that in the constructor, but it makes more sense to me to have that behavior tied to each execution rather than to the fuzzer itself.

Reply:
Ah I see, yes. You could probably work around it by resetting it. Jianwen's suggestion would make the fuzzer faster (by avoiding setting expectations at runtime), but depending on how fast the fuzzer is, it may or may not be necessary. How fast does it run?

Reply:
It currently runs at ~225 exec/s - I think the main performance bottleneck is setting up the mock expectations for `os_sys_calls_`. I do plan on eventually switching to a fake implementation (framework for that is already in place), but I'm waiting to see what the fuzz target for tls_inspector will require. Planning to revisit improving performance in a separate PR after both http_inspector and tls_inspector are covered.

Reply:
Revisiting for performance sounds fine to me.